### ISSUE TITLE
Really slow to create control

### ISSUE THREAD

**Issue by ConfusedVorlon** [jdstmporter/ColourWheel#1]
Hi,
I just came across your library. It's great and just what I need.

One wierd thing though; It takes about 3 seconds to create the control from a storyboard.
(on a fairly recent iPad air)

I'm working around this by just keeping a static version of my picker view controller - but it seems like an odd slowdown in a control that doesn't look like it should be super-complex

I'm using everything in with default settings

any suggestions?

thank you.

**Comment by jdstmporter** [comment]
Just checking.  I know I had this issue once or twice and am trying to jog my memory as to why.

**Comment by jdstmporter** [comment]
Could you send me some more details about your application, e.g. OS Version, Swift version, etc.  Also, it would be useful if you could send me part of your code, so I can see it in context?  Thank you.

**Comment by ConfusedVorlon** [comment]
thanks for the quick response

here is a project demonstrating the issue.
https://www.dropbox.com/s/7gs6mvnbx274z8n/Wheel.zip?dl=0

it's just a button that triggers loading and presenting a view controller (from storyboard)

I'm testing on a newish iPad air running 13.3
IOS target 13.2
Swift 5

**Comment by jdstmporter** [comment]
Thank you.  I'll check this out tomorrow.

**Comment by jdstmporter** [comment]
First of all, apologies for taking so long to respond.  I was ill.

Second, I'm actually having difficulty reproducing your problem, which is annoying, as obviously you can see it, but as I can't I'm not sure what to do.  However, I *think* it's something to do with the new-style SwiftUI constraints mechanism.

Looking at `ColourPickerVC.storyboard`, I note that it's currently set for the main `View` to use the new approach to layout (_Translates Mask into Constraints_), while the `Picker` uses the old approach (_auto_).  It tried switching them both to _auto_ and I started to get some second delay before the `Picker` first appeared.  Changing back, that delay seemed to go away.  

So, the problem may be simply that this is a pre-SwiftUI control, and I need to update it to take SwiftUI into account.  Would that be an acceptable next step?

**Comment by ConfusedVorlon** [comment]
Thanks for getting back on this - I appreciate this.

Do you mean SwiftUI? 

My demo project isn't using SwiftUI in any way. (It does have a couple of leftover import statements, but you can remove them as they're not actually used)

Constraints don't sound like a very likely problem. I'm manually fixing the height/width of the picker in the demo, so it seems unlikely that the solver would be having problems there...

Of course - a SwiftUI picker would be fantastic - but that's a separate thing :)

**Comment by jdstmporter** [comment]
I seem to have become confused.

The reason I suspect constraints is that, if you set them up correctly, with the picker a fixed size (as you have) then the (fairly intensive) process of building the colour wheel itself (which involves computing and setting the colour of a bitmap pixel by pixel to get really clean graphics) will be done once and for all in your Storyboard / NIBFile.   On the other hand, if the width of the thing is computed dynamically, it can take a very long time to do that once-off work.

Have a look at the example : the version on `master` uses a width based on the screen-size, and loading it on an iPad takes a *long time*.  On the other hand, the version on branch `fixed-width` fixes the width at 335 pixels, and it loads more or less instantly.

Given its similarity to that you're doing, the only thing I can think of that might be causing a bit of delay is that the width of the pop-over view you're using may be dynamically computed.  What happens if you replace it with a full-screen view, like mine?

**Comment by ConfusedVorlon** [comment]
That's really interesting. 

if I turn off the presentation animation, then:

presenting fullscreen, the controller takes ~ 1.6seconds
presenting as a form, it takes 3.3, so about twice as long

digging down, if the view controller is a form, then layoutSubviews is called twice on ColorWheel
so updateImage is called twice.
I assume that this is the expensive function.

the frame hasn't changed between the two layout calls, a check on whether the frame has changed saves some time

(now 1.9 seconds)

```
    var laidOutFrame:CGRect?
    public override func layoutSubviews() {
        print("layout: \(self.frame)")
        super.layoutSubviews()
        
        if self.frame != laidOutFrame {
            mid=bounds.mid
            radius=0.5*Swift.min(bounds.width,bounds.height)-Swift.max(0,borderWidth)
            updateImage()
            laidOutFrame = self.frame
        }
    }
```

But here's the weirdest Heisenbuggyness. If I run the code on the same iPad with the time profiler, then showing the controller takes 0.4 seconds.

This seems nuts - but it's really true. (testing on my iPad Air 2 (Model A1566))

I tried debugging with release-level optimisation turned on, and I don't see a speed change, so I don't think that's the issue.

It's very odd.

4x speed gain by turning on the time profiler (which should slow things down...)

**Comment by jdstmporter** [comment]
That is all seriously weird.  

You're right in identifying the slow bit.  I can see a possible work-around for that: if I do some cacheing / checking, so if the size hasn't changed since the last layout, it doesn't bother to recompute, then that should work.  I'll try that tomorrow.

As for why it does a double layout - thanks for that indicator.  I'll have a dig to see what the triggers are.  Might not got anywhere, but surely interesting to know.

And as the four-fold speed up when you're running Instruments?  Er, WHAT?  That shouldn't happen.

So I have two actions:

1. Put in checks / a cache so there control doesn't rebuild the wheel if it already has one the right size
2. Check calling conditions for the layout

I'll start in on (1) tomorrow.

**Comment by ConfusedVorlon** [comment]
awesome - thank you.

**Comment by jdstmporter** [comment]
I have implemented a simple kludge that means that, though the layout method is still called twice (I don't seem to be able to do anything about the - presumably something weird deep in the guts of UIKit) - it only does the work of computing the wheel first time round.

It seems a bit faster, but I have yet to do a proper timing test.  SO, before I release it, I just need to :

(1) check whether I have introduced any unanticipated new features
(2) hopefully come up with a slightly more elegant way of doing it
(3) rebuild the Pod and issue a new version

If you'd like to try out a version of it, I'm happy to oblige, but be warned, it isn't nicely packaged as a Pod (obviously) and I don't guarantee anything much by way of not doing weird, unexpected things.

**Comment by ConfusedVorlon** [comment]
thanks - I'm running from that branch now and it seems to be fine.

it certainly looks like you have stopped the double calculation

thanks.

I'd love to know why the profiler speeds things up. That feels like a compiler bug.

**Comment by ConfusedVorlon** [comment]
ok - the profiling vs debug thing turns out not to be mysterious at all.
profiling uses release configuration.
If I switch the run settings to use the release config, then I get the same timing without the profiler.

which is now a very manageable 0.5 seconds.

**Comment by jdstmporter** [comment]
Thank you for working that out.  Of course, it begs the question of _why_ debug settings caused a double render, but, as I said above, I suspect that's something deep in the implementation of Storyboards that we'll never know.

I won't close the issue just yet, as I haven't quite finished testing, but as soon as I have, I will update the Pod and then close the issue.

**Comment by ConfusedVorlon** [comment]
to be clear - the double layout thing is true in debug and release.

It's just that in release, the work is much faster generating the control.

removing the double layout reduces the time to display from ~0.51s to ~0.45s
it's not a lot, but it isn't nothing either!

**Comment by jdstmporter** [comment]
Good.  Thank you.  Issue closed, I think, with new version 5.0.1.
